### ISSUE TITLE
InexactError() constructing PolynomialRatio

### ISSUE THREAD

**Issue by ma-laforge** [JuliaDSP/DSP.jl#142]
The following throws an `InexactError()`:
```
using DSP
b=[1]; a=[10]
PolynomialRatio(b,a)
```

which can somehow be avoided by making `a` an array of `Float64`, instead:
```
using DSP
b=[1]; a=[10.0]
PolynomialRatio(b,a)
```

Also: there is no error if `b=[10]; a=[1]`, for some reason.

**Comment by hamzaelsaawy** [comment]
Its from [here](https://github.com/JuliaDSP/DSP.jl/blob/73cfda49019143fe6ed1b6a72555778470f16e34/src/Filters/coefficients.jl#L53), where the numerator and denominator are divided by the leading term of the denominator polynomial.

Is there a need to do so?

**Comment by ma-laforge** [comment]
No.  There is no real need to do so.  It was just a little unexpected.

I think from the ***point of view of usability***, it is quite confusing to see an inexact error when all you do is construct a polynomial with 2 arrays of integers.

However, I do find it a bit odd that the polynomials get normalized when they get created.  That is also unexpected.  I personally would not have done this normalization unless there was a really good reason for it (which there might be, I suppose).

PS Thanks for the feedback.

**Comment by martinholters** [comment]
We could add an optional parameter to `PolynomialRatio` to suppress normalization, defaulting to the current behavior for compatibility.

**Comment by ma-laforge** [comment]
I don't really have a strong opinion about the optional parameter.

It would not help with the confusing message - because you would still need to be aware that the default behaviour is to normalize, so the problem would remain.

But this brings up an interesting question: why ***does*** DSP.jl normalize the polynomial at all if it is still just as valid ***not*** to do the normalization?

Shouldn't `normalize()` be a function that is called by the functions operating on the polynomial (assuming that normalization is actually needed for the operation)?

**Comment by hamzaelsaawy** [comment]
I think normalizing is used primarily in the  [FIR filtering algorithm](https://github.com/JuliaDSP/DSP.jl/blob/5abc50754793fc6cece466454e5330839125aa65/src/dspbase.jl#L76) but interestingly enough that code doesn't assume the coefficients are normalized, so maybe we can do without in the `PolynomialRatio` object?  

The only issue would be a `PolynomialRatio{Int}` might return a `Biquad{Float64}` since the latter is normalized by definition, and that would be weird for some people to see, but I think a lot of the filter coefficient objects change types when the representation changes.

**Comment by ma-laforge** [comment]
@hamzaelsaawy:
I do not completely understand all the details of what you are saying, but the overall idea seems good.

### Small disagreement
I think `filt(::PolynomialRatio{T}, ::Vector{Int})`, should probably return a `Vector{Int}` - irrespective of type `T`.  If you are working with discrete levels, the fact that the filter is made up of Real numbers or Integer values should not change the fact that the result should also have discrete levels.

### Type stability
... Of course, the `normalize()` function (or `normalize=true` argument ***could be made*** to generate `PolynomialRatio{Float64}`  values irrespective of the value of `a[1]`.  This seems like a perfectly valid solution that avoids type instabilities.

**Comment by hamzaelsaawy** [comment]
I think biting the bullet and adding a `normalize=false` keyword is the best solution.
I am changing how the code handles types for the coefficients/design parts, and this fits nicely.

I honestly hadn't thought about filtering integer streams.... maybe add a parameter to set the output type of `filt`?
I am concerned about how often I jam in vectors of integers because I am too lazy to type a period to get an array of floats, so I would vote to bias towards `Float64`, maybe with `filt(::Type{T}, ::FilterCoefficient, ::Vector)` returning a `Vector{T}` so unless the user actively is looking to perform integer filtering.
This ties up type stability nicely too.

But it does bring up a related question about complex coefficients since currently `PolynomialRatio(ZeroPoleGain([1im], [0], 1))` gives `x/(x-1)`  and not `(x-im)/(x-1)` since the code assumes you are always working in the real domain and dumps the imaginary part of the coefficients.

**Comment by ssfrr** [comment]
> I think filt(::PolynomialRatio{T}, ::Vector{Int}), should probably return a Vector{Int} - irrespective of type T.

It seems like a good general rule that combining different element types gives a result that follows the normal promotion rules, e.g. multiplying an `Int` by a `Float64` gives a `Float64`. Special-casing some operations seems like it adds more complexity and would need a very good justification.

**Comment by martinholters** [comment]
I'm also not quite clear what should happen for `filt(::PolynomialRatio{Int}, ::Vector{Int})` when the filter has not been normalized (assuming the expected element type of the result is `Int`). Should it just error (`InexactError`)? Do the necessary division as in integer division? Or what?

**Comment by hamzaelsaawy** [comment]
My two cents are that since (digital) filtering is addition, multiplication, and division, we follow their promotion rules, specifically the later:
`/(::Int, ::Int) :: Float64`, so maybe `filt(::PolynomialRatio{Int}, ::Vector{Int}) :: Vector{Float64}`

Currently, `filt(::PolynomialRatio{Int}, ::Vector{Int}) :: Vector{Int}`, which means changing would break compatibility, but I think the `filt([::Type{T}=Float64], ::FilterCoefficient, ::Vector) :: Vector{T}` could be a way to keep that functionality, and then automatically call `round`/`real` based on `T`.

**Comment by ma-laforge** [comment]
My personal belief is that `filt()` should ***not*** adhere to the promotion rules.  `filt()` is ***not*** a simple mathematical operation that needs promotion rules to maintain accuracy.

I would argue that we look to the applications to discern how it should be used.

### Case 1: Analog signal with "real" filter

Similarly, take the example of a ***physical*** analog filter made up of RLC components and op-amps:
  - A very real-life typical example.
  - Processed by a continuous-time filter having complex-conjugate poles/zeros (in the general case).  Note that the ***polynomial*** has real coefficients, though.
  - The filter output will be real-valued as well (not complex).

The ouput vector type should match that of the input signal vector.
  - `filt(coeff::PolynomialRatio{TP<:Real}, sig::Vector{T}) => Vector{T}`
  - ex: `filt(::PolynomialRatio{Float64}, ::Vector{Float64}) => Vector{Float64}`
  - ex: `filt(::PolynomialRatio{Int64}, ::Vector{Float64}) => Vector{Float64}`

I really do not see how the element type of `PolynomialRatio` should affect the output vector type (unless it is `<:Complex` - see "Case 3").

### Case 2: "Digital signal" with "real" filter

Here we want to filter a "digital" signal with discrete-levels (ex: an audio PCM signal stored with integer values).
  - A very real-life, typical example.
  - Processed by a filter having complex-conjugate poles/zeros (in the general case).  Again, the ***polynomial*** has real coefficients here.
  - In general, the filter output will be real-valued (not complex), and typical applications would want to maintain level discretization.

I propose that the resultant filtered signal should therefore remain a vector of integer values, for example:
  - `filt(::PolynomialRatio{Float64}, ::Vector{Int64}) => Vector{Int64}`

If I desired a `Vector{Float64}` output, I would personally expect to feed the algorithm a `Vector{Float64}`:
```
signal = [1,2] #Some signal
signalF64 = Vector{Float64}(signal) #Explicitly convert to a Float64
filt(filtercoeff, signalF64) #Returns Vector{Float64}
```

Which would make use of the floating point "version":
  - `filt(::PolynomialRatio{<:Real}, ::Vector{Float64}) => Vector{Float64}`

### Case 3: "Digital signal" with "complex" filter

Here we are doing a more fancy DSP operation where the filter output is ***complex*** (not real).
  - Processed by a filter that is ***no longer limited*** to complex-conjugate poles/zeros.  In this case, the ***polynomial*** has complex coefficients in the general case.
  - In general, the filter output will be complex - so ***type promotion would likely be needed in this case***.

I would therefore propose that the resultant filtered signal should be promoted to a `Complex` of whatever data type was used in the input signal vector:
  - `filt(::PolynomialRatio{Complex{TP<:Real}}, ::Vector{T}) => Vector{Complex{T}}`
  - ex: `filt(::PolynomialRatio{Complex{Float64}}, ::Vector{Int64}) => Vector{Complex{Int64}}`
  - ex: `filt(::PolynomialRatio{Complex{Int64}}, ::Vector{Float64}) => Vector{Complex{Float64}}`

### Comments

  - The fact that we normalize the coefficients of a polynomial does not actually affect what it represents.
  - The fact that we use integer values or real values to store polynomials does not affect what it represents either (unless there are rounding errors).
  - I therefore see no reason for the coefficients of a `PolynomialRatio` to be converted/post-processed... Other than for something like efficiency of the implementation.
  - The discrete nature (integer vs real) of the polynomial coefficients has no direct impact on the output signal type (integer or real).
  - Only the "real vs complex" nature of the polynomial coefficients has an impact.  Complex polynomial coefficients convert (in the general case) real-valued signals to complex-valued signals.

I mostly think the statements I made here are correct.  I do not believe I made a significant mistake.

**Comment by galenlynch** [comment]
When I filter a fixed point / integer signal with Julia, I seldom want a fixed point result. Doing so just adds unnecessary quantization noise. While analog to digital converters  / online DSPs may be constrained to using fixed points arithmetic, I run Julia on a computer that is not. Floating point results are more accurate, and therefore seem strictly better to me. If, for some reason, I need the output to be fixed point, then I can round it myself. 

Giving the user the most accurate result seems like sensible behavior, to me.

**Comment by ma-laforge** [comment]
> @galenlynch: When I filter a fixed point / integer signal with Julia, I seldom want a fixed point result.

I could imagine this to be a typical application as well.

However, I would personally expect that in a case like this, the input vector should be cast to a vector of floating points using a simple call to: `Vector{Float64}(input_vector)`.

My main arguments are:

  - I personally feel that Julia functions try to maintain the input data type for its output ***unless*** significant inaccuracies would occur (in which case, promotion rules apply - as hilighted by @ssfrr).
  - There are probably filter implementations that are more efficient/fast (yet still accurate) when executed in the fixed-point domain.  And I don't particularly like having to call a specially named function to execute that algorithm.  It seems to diverge from Julia's typical multiple dispatch philosophy.

Having said that, I must agree with @hamzaelsaawy:  adding an explicit output type to the `filt()` function is probably the most flexible solution (while maintaining a Julian feel):

```
filt([otype::Type{T}=Float64], ::PolynomialRatio, ::Vector) :: Vector{T}
```

...because it allows for a wider variety of optimized implementations.  For example, you can use `otype` to ***explicitly*** request a `filt()` operation that transforms `Vector{Int64}->Vector{Int64}`, or one that transforms `Vector{Int64}->Vector{Float64}`.

**Comment by galenlynch** [comment]
Rounding to maintain integer (or other fixed point) output would necessarily lead to inaccuracies, so I think sticking with standard promotion rules is a sensible default. However, I'm not opposed to adding a way to specify the output type, as long as it's backwards compatible.

The original issue seems like something that should be fixed with `PolynomialRatio`, at the very least by throwing a more informative error. I honestly don't know why the filter coefficients are normalized, I'm more familiar with the FIR filtering in `filt.jl` than the IIR filtering in `dspbase.jl`.

**Comment by hamzaelsaawy** [comment]
I would like to point out that Julia throws an error when converting to an `Int` from a `Float` without an explicit `round`/`ceil`/`floor`/etc, or to a `<:Real` from a `Complex` so  an **implicit** round/truncation when using `filt(otype::Type, ...)` may be not be Julia-esque. 
(The current functionality throws an `InexactError` for `filt!(out::Vector{Int}, filter<:FilterCoefficients, x::Vector)` if the output is actually Floating)

If we are happy with that, I am happy to start adding an output type to `filt` and de-normalize `PolynomialRatios`. 
Maybe we can add a warning banner in the readme/docs...

**Comment by ma-laforge** [comment]
> @galenlynch: Floating point results are more accurate, and therefore seem strictly better to me.

That statement is partially correct because many ***operations*** we typically perform will tend to yield more accurate results if we use floating point values.

And, though I am sure you already know this, the above statement is technically inaccurate:  Floating point values provide you with more ***dynamic range***;  They do not result in higher accuracy.  In fact, floating point values technically ***sacrifice accuracy*** by however number of bits are allocated to the mantissa (exponent part).

Having said that, I do realize that ***we all*** tend to make the over-simplified statement that "floating point results are more accurate" - as a matter of habit.

### Counterpoint: Accumulation errors

As a side note, I am pretty certain that filters that perform integrations will tend to be ***less accurate*** with floating point operations than with fixed point operations.  Floating point algorithms are highly succeptible to accumulation errors that are not present in fixed-point arithmetic (provided that the *accumulated quantities* are exactly represented by an integer value).

**Comment by ma-laforge** [comment]
> @hamzaelsaawy: [...] so an **implicit** round/truncation when using `filt(otype::Type, ...)` may be not be Julia-esque.

I could accept that statement if we were to ommit the `otype` argument.  But technically: by specifying `otype`, we would now have an ***explicit*** round/truncation operation (since the user explicitly requested it).

### Specialized algorithms

In actuality, I was not necessarily talking about creating a function that rounded the results before returning the output vector.  I was actually considering the implementation of specialized functions that are faster/more accurate/etc than the generic floating-point version.  This type of function overloading is ***very much*** Julia-esque.

For example, I am ~95% certain that a large class of filtering applications in advanced DSP chips (FPGAs/ASICs) used in real-life products are implemented with ***fixed-point algorithms***!

And, if I am not mistaken, these algorithms do not suffer from rounding errors other than the initial rounding of the coefficients and input data stream.

In reality, the errors to watch out for are more likely to be overflow errors in the intermediate values.  In other words, you might have to use `Int128` or something similar to do the intermediate computations on an input vector of `Int64`-s.

**Comment by ssfrr** [comment]
Given that reasonable and intelligent people can have diverging opinions on what they would want, I think it's best to default to making as few assumptions as possible and provide simple functionality that composes well. Given that type promotion is already pretty well-defined and widely-used in Julia, I'd advocate for sticking to that, and having `eltype(foo(a, b)) == promote_type(eltype(a), eltype(b))` whenever possible.

It seems like the crux of this issue is the normalization that's automatically performed on `PolynomialRatio`, which takes an operation that would otherwise be `*` and `+` and introduces a pesky `/`.

Ultimately filtering is computing each output sample as a linear function of the input and output history. If the input and the multiplication coefficients are all integer-valued, there shouldn't be any reason to convert the output to floats (which as @ma-laforge points out is actually _less_ accurate).

### Minimal Change:

One minimal change to make would be to add a `normalize=true` kwarg to `PolynomialRatio` (as @martinholters [suggested](https://github.com/JuliaDSP/DSP.jl/issues/142#issuecomment-462263845)), and throw a more useful error in the case that the division would give a non-integer number. (perhaps the error could suggest setting `normalize=false` in this case). This could be as simple as adding a `try...catch` that catches the `InexactError` rethrows a more useful error.

### slightly less minimal change

do the above, but also throw a deprecation if `normalize` is unspecified, so that we can change the behavior in a future release (and possibly remove the kwarg). It doesn't seem like anyone has a particularly compelling reason to normalize by default, so IMO it would be better not to, and perhaps add a `normalize(p::PolynomialRatio)` such that `normalize(p::PolynomialRatio{Int}) isa PolynomialRatio{Float64}`.

**Comment by ma-laforge** [comment]
@ssfrr :

  - Agree with minimal change.  If we stick with this behaviour, having a more descriptive error would be nice.
  - No strong opinion on the deprecation thing.  I have a very poor undertanding on how to deprecate things properly - especially how one can make their way to a more desireable API without breaking things.
  - Agree with `normalize(p::PolynomialRatio)` such that `normalize(p::PolynomialRatio{Int}) isa PolynomialRatio{Float64}`.

Even though I don't have a strong opinion on how to deprecate, I do like the overall proposal from @ssfrr.

**Comment by galenlynch** [comment]
I would pump the breaks on making any changes to normalization before there's some sort of review of the IIR algorithms used in `dspbase.jl` and elsewhere. I didn't write any of that code, and don't claim to understand the assumptions that it's made, but it seems like normalization was present in the very first commit with filtering functionality (8ce5a64e0c29f5acdd0fcc33477b7d2797a96007). 

From the conversation here it seems like there's little certainty either way over whether this normalization is necessary (read: assumed by subsequent functions). Brief googling makes me think that simple difference equations often require/assume normalized coefficients, e.g. [here](https://en.wikipedia.org/wiki/Infinite_impulse_response#Transfer_function_derivation) or [here](http://www.eas.uccs.edu/~mwickert/ece2610/lecture_notes/ece2610_chap8.pdf). Looking at `_filt_iir!` in `dspbase.jl`, it looks like `a[1]` is never touched, and therefore most likely the author assumed it to be `1`. This makes sense, as normalizing the filter coefficients should let you get rid of one multiplication for every output sample, which is well worth the up front cost of normalization.

If this is the case and normalization is indeed necessary, then maybe a simple solution would be an inner constructor for `PolynomialRatio{<:Integer}` that enforces `a[end] == 1` to avoid InexactErrors.

**Comment by galenlynch** [comment]
Note that the coefficients are reversed by `coefa` and `coefb`, so `a[1]` during filtering is `a[end]` when constructing `PolynomialRatio`.

**Comment by galenlynch** [comment]
Oh I see, `filt!` checks for normalization and enforces it. So all this discussion about type conversions is moot when it comes to filtering with `PolynomialRatio`? It seems like both `a` and `b` will be converted to floats if `PolynomialRatio` is not already normalized.

Why not just enforce normalization at construction, as is currently the case? Removing it form `PolynomialRatio` will just move the issue around. Furthermore, there's a real performance benefit to enforcing normalization.

**Comment by galenlynch** [comment]
Let me be a little bit more clear about what I propose:

1) Keep normalization in `PolynomialRatio` for performance reasons, and because downstream functions will just normalize the coefficients anyways.
2) Have an outer constructor for `PolynomialRatio` that will convert integer coefficients to floats and normalize them, as this is reasonable behavior that I think most users wouldn't be that surprised by.
3) If a user REALLY wants integer coefficients, they can construct a `PolynomialRatio{Int}` using the type parameter, and deal with a persnickety inner constructor that enforces normalization, and offers useful errors (i.e. ask the user to supply already normalized coefficients). 

The fact is that constructing `PolynomialRatio` with integers with the current code must have required users to already normalize the coefficients, or get very lucky, to avoid this `InexactError`.

**Comment by martinholters** [comment]
When filtering with a `PolynomialRatio`, one has to divide by a_0 one way or another---either upon construction of the `PolynomialRatio` or during filtering itself. If we do normalize upon construction, filtering only involves multiplication and addition and it becomes natural for `filt(::PolynomialRatio{Int}, ::Vector{Int})` to return a `Vector{Int}`.
If we do not normalize, I see the following options for `filt(::PolynomialRatio{Int}, ::Vector{Int})`:
* Promote to `Float64`. Doesn't allow to give a `Vector{Int}`.
* Only promote to `Float64` if a_0 isn't one. That would mean type-instability.
* Always return a `Vector{Int}`
  - rounding (truncating?) the result as necessary. Very un-Julian.
  - throwing an `InexactError` if a_0 isn't one.

I don't especially like any of these. That leaves normalization upon construction. Here I very much like the proposal by @galenlynch: Always normalize, promoting the type as needed. If the eltype is given explicitly (`PolynomialRatio{Int}`), don't promote (obviously), if normalization yields values not `convert`ible to the given eltype, error. I think that is my favorite solution here.

An alternative could be to distinguish normalized and non-normalized polynomial ratios on the type level (`NormalizedPolynomialRatio` vs. `NonNormalizedPolynomialRatio` or whatever), but that looks like overkill.

A final remark: It would be really nice if whatever we do here just works in conjunction with FixedPointNumbers.jl. If have no idea to what extent that is the case already.

**Comment by martinholters** [comment]
The original issue is fixed by #284 now:
```julia
ulia> b=[1]; a=[10];

julia> PolynomialRatio(b,a)
PolynomialRatio{:z, Float64}(Polynomials.LaurentPolynomial(0.1), Polynomials.LaurentPolynomial(1.0))
```
Note the propagation to `Float64` happening automatically. Even not required to avoid type instability:
```julia
julia> PolynomialRatio([10], [1])
PolynomialRatio{:z, Float64}(Polynomials.LaurentPolynomial(10.0), Polynomials.LaurentPolynomial(1.0))
```
To obtain a `PolynomialRatio{:z, Int}`, construct one:
```julia
julia> PolynomialRatio{:z, Int}([10], [1])
PolynomialRatio{:z, Int64}(Polynomials.LaurentPolynomial(10), Polynomials.LaurentPolynomial(1))
```
But that will fail of normalization to a0=1 would give a non-integer coefficient:
```julia
julia> PolynomialRatio{:z, Int}(b, a)
ERROR: InexactError: Int64(0.1)
```
This gives control over the coefficient type and at the same time ensures a0=1. If we really feel that we need to store coefficients with a0≠1, we can always add a dedicated type (or maybe a type parameter to `PolynomialRatio`) later.
